### PR TITLE
[FLINK-30332] HsFileDataIndex supports caching and spilling index entry

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*`HsFileDataIndex` is maintains in a list of treeMap(i.e. in memory), there is no problem in our 10T TPC-DS test. However, with the increase of job's data volume, the index will continue to grow, and there is a risk of OOM. We should support index data spilling and loading to avoid this problem.*

## Brief change log

  - *HsFileDataIndex supports caching and spilling index entry*

## Verifying this change

This change added unit tests .

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 1]
```diff
@@ -270,6 +270,24 @@ public class NettyShuffleEnvironmentOptions {
...
+                    .defaultValue(256)
+                    .withDescription(
+                            "Controls the segment size of hybrid spilled file data index.");
+
+    /** Max number of hybrid retained regions in memory. */
+    @Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
+    public static final ConfigOption<Long> HYBRID_SHUFFLE_NUM_RETAINED_IN_MEMORY_REGIONS_MAX =
+            key("taskmanager.network.hybrid-shuffle.num-retained-in-memory-regions-max")
+                    .longType()
+                    .defaultValue(Long.MAX_VALUE)
```
Comment (xintongsong) [inline flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:287]: This practically means the feature is disabled. I'm not sure how many users will be aware of this knob.

File: flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java [hunk 2]
```diff
@@ -270,6 +270,24 @@ public class NettyShuffleEnvironmentOptions {
...
                                     "taskmanager.memory.framework.off-heap.batch-shuffle.size"));
 
+    /** Segment size of hybrid spilled file data index. */
+    @Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER_NETWORK)
+    public static final ConfigOption<Integer> HYBRID_SHUFFLE_SPILLED_INDEX_SEGMENT_SIZE =
+            key("taskmanager.network.hybrid-shuffle.spill-index-segment-size")
+                    .intType()
+                    .defaultValue(256)
+                    .withDescription(
+                            "Controls the segment size of hybrid spilled file data index.");
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-core/src/main/java/org/apache/flink/configuration/NettyShuffleEnvironmentOptions.java:280]: Add description of unit and change this default value to 1024.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataIndexSpilledRegionManagerImpl.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+ *   <li>Multiple regions belonging to the same subpartition form a segment.
+ *   <li>The regions in the same segment have no special relationship, but are only related to the
+ *       order in which they are spilled.
+ *   <li>Each segment is independent. Even if the previous segment is not full, the next segment can
+ *       still be allocated.
+ *   <li>If a region has been written to the index file already, spill it again will overwrite the
+ *       previous region.
+ *   <li>The very large region will monopolize a single segment.
+ * </ul>
+ */
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataIndexSpilledRegionManagerImpl.java:52]: 1.By design, there is no concept of segment in the interface `HsFileDataIndexSpilledRegionManager `. The protocol only concerns about the spilling of region.
2.Added it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataIndexSpilledRegionManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+ *        ||              Big Region             |              |
+ *        | - - - - - - - - - - - - - - - - - - -               |
+ *         - - - - - - - - - - - - - - - - - - - - - -- - - - -
+ * </pre>
+ */
+public class HsFileDataIndexSpilledRegionManagerImpl
+        implements HsFileDataIndexSpilledRegionManager {
+
+    /** Reusable buffer used to read and write the immutable part of region. */
+    private final ByteBuffer immutablePartBuffer =
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsFileDataIndexSpilledRegionManagerImpl.java:74]:
```suggestion
    private final ByteBuffer regionHeaderBuffer =
```

**Review by xintongsong — APPROVED** [review]
Thanks, @reswqa. LGTM. Nice work~!

I'm merging this once the CI passes. There's only one minor comment, which I'll address myself during the merging.

**Comment by reswqa** [comment]
Thanks @xintongsong for the review and fix-up~

**Pull request closed without merging**